### ISSUE TITLE
Incorrect `unresolved-reference` for variables bound in generators in a condition

### ISSUE THREAD

**Issue by alexchandel** [astral-sh/ty#3599]
### Summary

In the following snippet, `char` is guaranteed to be bound because `any` is trivially false for empty sets, and further by static analysis we can see the generator yields 16 times (one for each character in the value `'*."/\\[]:;|,<>\r\n\t'`), meaning `char` will certainly be bound.

```py
if any((char := c) in name for c in '*."/\\[]:;|,<>\r\n\t'):
    raise RuntimeError(f"Invalid character {char!r}")
```

Yet ty reports `error[unresolved-reference]: Name `char` used when not defined`

### Version

ty 0.0.40 (7b95bc219 2026-05-27)

**Comment by AlexWaygood** [comment]
Thanks! This is https://github.com/astral-sh/ty/issues/162
